resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/CircuitBreakerConfig.java

```diff
@@ -251,6 +251,7 @@ public String toString() {
             .ofSeconds(DEFAULT_SLOW_CALL_DURATION_THRESHOLD);
         private Duration maxWaitDurationInHalfOpenState = Duration
             .ofSeconds(DEFAULT_WAIT_DURATION_IN_HALF_OPEN_STATE);
+        private byte setWaitInOpenStateCounter = 0;
```

Comment:
How about use `Duration waitDurationInOpenState` field and check like `if (waitDurationInOpenState != null && waitIntervalFunctionInOpenState != null)` at `build()`?

Reply:
Maybe i move initialization  `waitIntervalFunctionInOpenState` in `builder()` and rename `validateWaitIntervalFunctionInOpenState` to `createWaitIntervalFunctionInOpenState`?

```java
 private IntervalFunction createWaitIntervalFunctionInOpenState() {
            if (waitDurationInOpenState != null && waitIntervalFunctionInOpenState != null ) {
                throw new IllegalStateException("The waitIntervalFunction was configured twice which could result in an undesired state. " +
                    "Please use either waitIntervalFunctionInOpenState or waitDurationInOpenState.");
            }
            return waitDurationInOpenState != null ? IntervalFunction.of(waitDurationInOpenState.toMillis()) : waitIntervalFunctionInOpenState;
        }
```

Reply:
Yes, seems fine to me.

Reply:
Yes exactly. what do you think @RobWin ?

Reply:
Yes exactly.

Reply:
Sorry, i have to clarify one more solution
I want use ` private long waitDurationInMillis = -1L;` instead of `Duration waitDurationInOpenState`.
What would will not calculate `waitDurationInOpenState.toMillis()` twice. 
Because in `public Builder waitDurationInOpenState(Duration waitDurationInOpenState) ` need validation `waitDurationInMillis < 1` what would rise throw developer can saw where method throw exception  in stack trace.
```java
private IntervalFunction createWaitIntervalFunctionInOpenState() {
            final boolean wasItSetToWaitDurationInMillis = waitDurationInMillis > 0;
            if (wasItSetToWaitDurationInMillis && waitIntervalFunctionInOpenState != null ) {
                throw new IllegalStateException("The waitIntervalFunction was configured twice which could result in an undesired state. " +
                    "Please use either waitIntervalFunctionInOpenState or waitDurationInOpenState.");
            }
            return wasItSetToWaitDurationInMillis ? IntervalFunction.of(waitDurationInMillis) : waitIntervalFunctionInOpenState;
        }

```

Reply:
I think this optimization is not needed. Shouldn't be performance critical.

Reply:
Agree with Robert.

Reply:
one more clarification, again) sorry
Could remind you that `waitIntervalFunctionInOpenState` has default value.
And if we use `if (wasItSetToWaitDurationInMillis && waitIntervalFunctionInOpenState != null ) ` we should remove default value for `waitIntervalFunctionInOpenState`.

My attempt
```java
        private IntervalFunction createWaitIntervalFunctionInOpenState() {
            if (waitDurationInOpenState != null && waitIntervalFunctionInOpenState != null) {
                throw new IllegalStateException("The waitIntervalFunction was configured twice which could result in an undesired state. " +
                    "Please use either waitIntervalFunctionInOpenState or waitDurationInOpenState.");
            }

            if (waitDurationInOpenState != null) {
                return IntervalFunction.of(waitDurationInOpenState.toMillis());
            }

            return waitIntervalFunctionInOpenState == null
                ? IntervalFunction.of(Duration.ofSeconds(DEFAULT_SLOW_CALL_DURATION_THRESHOLD))
                : waitIntervalFunctionInOpenState;
        }
```

But, when we use `CircuitBreakerConfig.from(CircuitBreakerConfig config)`

```java
        CircuitBreakerConfig baseConfig = custom()
            .waitDurationInOpenState(Duration.ofSeconds(100))
            .slidingWindowSize(1000)
            .permittedNumberOfCallsInHalfOpenState(100)
            .writableStackTraceEnabled(false)
            .automaticTransitionFromOpenToHalfOpenEnabled(true)
            .failureRateThreshold(20f).build();

        CircuitBreakerConfig extendedConfig = from(baseConfig)
            .waitDurationInOpenState(Duration.ofSeconds(20))
            .build();
```
We get Fail. If we remove default value for `waitIntervalFunctionInOpenState` we can loose backwards-compatibility.

I could try something else for the proposed solution... What you think?

Reply:
Yes, we must not loose  backwards-compatibility

Reply:
Maybe we just go back and use an int counter. If the counter == 2, then the user has used both methods.

Reply:
How about two boolean values like `isWaitDurationInOpenStateSet` and `isWaitIntervalFunctionInOpenStateSet`?

Reply:
Counter seems simpler to me :)

Reply:
Maybe we add a waitIntervalBiFunction in the future as well, comparable to Retry.

Reply:
I just worried about set same value twice like following. This is not a normal case.
```
CircuitBreakerConfig.custom()
    .waitDurationInOpenState(Duration.ofSeconds(100))
    .waitDurationInOpenState(Duration.ofSeconds(200))
    .build()
```

If it is ok, counter is also fine to me.

Reply:
Ok, I think we should not check this, because this could be done on any config property. If he overrides the same property, it's the fault of the user ;)

Reply:
Please could you consider combining the two solutions together:

```java
  private IntervalFunction createWaitIntervalFunctionInOpenState() {
            if (setWaitInOpenStateCounter > 1) {
                if (waitDurationInOpenState != null && waitIntervalFunctionInOpenState != null) {
                    throw new IllegalStateException("The waitIntervalFunction was configured twice which could result in an undesired state. " +
                        "Please use either waitIntervalFunctionInOpenState or waitDurationInOpenState.");
                }
            }
            if (waitDurationInOpenState != null) {
                return IntervalFunction.of(waitDurationInOpenState.toMillis());
            }

            return waitIntervalFunctionInOpenState == null
                ? IntervalFunction.of(Duration.ofSeconds(DEFAULT_SLOW_CALL_DURATION_THRESHOLD))
                : waitIntervalFunctionInOpenState;
        }
```
First `setWaitInOpenStateCounter > 1` avoid overridden Second `waitDurationInOpenState != null && waitIntervalFunctionInOpenState != null` avoid case which described @dlsrb6342

Reply:
oh, no my bad(

Reply:
Let's keep it simple. 
`waitIntervalFunctionInOpenState` has a default value.
It can be overwritten, but we check a counter and if counter > 1, then throw an exception in the build method.

Reply:
@RobWin let's focus on the solution with equal 2 or greater than 1?

Reply:
Greater than 1 is fine. 
We can change it to `The waitIntervalFunction was configured multime times which could result in an undesired state.`

Reply:
@RobWin what did you mean `multime`? Maybe you mean `multiple`?

Reply:
Yes, sorry ;)

Reply:
Thanks